File: UsingFMDB-Objective-C/UsingFMDB-Objective-C/BookCache.m
//
//  BookCache.m
//  UsingFMDB-Objective-C
//
//  Created by akabeko on 2016/12/26.
//  Copyright © 2016年 akabeko. All rights reserved.
//

#import "BookCache.h"
#import "Book.h"

@interface BookCache ()

/** Collection of author names. */
@property (nonatomic, readwrite) NSMutableArray *authors;

/** Dictionary of book collection classified by author name. */
@property (nonatomic, readwrite) NSMutableDictionary *booksByAuthor;

@end

@implementation BookCache

/**
 * Initialize the instance.
 *
 * @return Instance.
 */
- (instancetype)init {
    self = [super init];
    if (self) {
        self.authors       = [NSMutableArray array];
        self.booksByAuthor = [NSMutableDictionary dictionary];
    }

    return self;
}

/**
 * Initialize the instance.
 *
 * @param books Collection of the book data.
 *
 * @return Instance.
 */
- (instancetype)initWithBooks:(NSArray *)books {
    self = [self init];
    if (self) {
        for (NSInteger i = 0, max = [books count]; i < max; ++i) {
            [self add:[books objectAtIndex:i]];
        }
    }

    return self;
}

/**
 * Add the new book.
 *
 * @param book Book data.
 *
 * @return YES if successful.
 */
- (BOOL)add:(Book *)book {
    if (book.bookId == kBookIdNone) {
        return NO;
    }

    NSMutableArray *books = [self.booksByAuthor objectForKey:book.author];
    if (books) {
        [books addObject:book];
    } else {
        books = [NSMutableArray array];
        [books addObject:book];
        [self.booksByAuthor setObject:books forKey:book.author];
        [self.authors addObject:book.author];
    }

    return YES;
}

/**
 * Remove the new book.
 *
 * @param book Book data.
 *
 * @return YES if successful.
 */
- (BOOL)remove:(Book *)book {
    NSMutableArray *books = [self.booksByAuthor objectForKey:book.author];
    if (!(books)) {
        return NO;
    }

    BOOL succeeded = NO;
    for (NSInteger i = 0, max = [books count]; i < max; ++i) {
        Book *existBook = [books objectAtIndex:i];
        if (existBook.bookId == book.bookId) {
            [books removeObjectAtIndex:i];
            succeeded = YES;
            break;
        }
    }

    if (succeeded && [books count] == 0) {
        succeeded = [self removeAuthor:book.author];
    }

    return succeeded;
}

/**
 * Update the new book.
 *
 * @param oldBook Old book data.
 * @param newBook New book data.
 *
 * @return YES if successful.
 */
- (BOOL)update:(Book *)oldBook newBook:(Book *)newBook {
    if ([oldBook.author compare:newBook.author] == NSOrderedSame) {
        return [self replaceBook:newBook];
    } else if ([self remove:oldBook]) {
        return [self add:newBook];
    }

    return NO;
}

#pragma mark - Private

/**
 * Remove the author at the inner collection.
 *
 * @param author Name of the author.
 *
 * @return YES if successful.
 */
- (BOOL)removeAuthor:(NSString *)author {
    [self.booksByAuthor removeObjectForKey:author];
    for (NSInteger i = 0, max = [self.authors count]; i < max; ++i) {
        NSString *existAuthor = [self.authors objectAtIndex:i];
        if ([existAuthor compare:author] == NSOrderedSame) {
            [self.authors removeObjectAtIndex:i];
            return YES;
        }
    }

    return NO;
}

/**
 * Replace the book.
 *
 * @param book Book data.
 *
 * @return YES if successful.
 */
- (BOOL)replaceBook:(Book *)book {
    NSMutableArray *books = [self.booksByAuthor objectForKey:book.author];
    if (!(books)) {
        return NO;
    }

    for (NSInteger i = 0, max = [books count]; i < max; ++i) {
        Book *existBook = [books objectAtIndex:i];
        if (existBook.bookId == book.bookId) {
            [books replaceObjectAtIndex:i withObject:book];
            return YES;
        }
    }

    return NO;
}

@end

File: UsingFMDB-Swift/UsingFMDB-Swift/BookCache.swift
//
//  BookCache.swift
//  UsingFMDB-Swift
//
//  Created by akabeko on 2016/12/22.
//  Copyright © 2016年 akabeko. All rights reserved.
//

import UIKit

/// Manager fot the book data.
class BookCache: NSObject {
    /// A collection of author names.
    var authors = Array<String>()

    /// Dictionary of book collection classified by author name.
    var booksByAuthor = Dictionary<String, Array<Book>>()

    /// Initialize the instance.
    override init() {
        super.init()
    }

    /// Initialize the instance.
    ///
    /// - Parameter books: Collection of the book data.
    init(books: Array<Book>) {
        super.init()

        books.forEach({ (book) in
            if !self.add(book: book) {
                print("Faild to add book: " + book.author + " - " + book.title )
            }
        })
    }

    /// Add the new book.
    ///
    /// - Parameter book: Book data.
    /// - Returns: "true" if successful.
    func add(book: Book) -> Bool {
        if book.bookId == Book.BookIdNone {
            return false
        }

        if var existBooks = self.booksByAuthor[book.author] {
            existBooks.append(book)
            self.booksByAuthor.updateValue(existBooks, forKey: book.author)
        } else {
            var newBooks = Array<Book>()
            newBooks.append(book)
            self.booksByAuthor[book.author] = newBooks
            self.authors.append(book.author)
        }

        return true
    }

    /// Remove the book.
    ///
    /// - Parameter book: Book data.
    /// - Returns: "true" if successful.
    func remove(book: Book) -> Bool {
        if var books = self.booksByAuthor[book.author] {
            for i in 0..<books.count {
                let existBook = books[i]
                if existBook.bookId == book.bookId {
                    books.remove(at: i)
                    self.booksByAuthor.updateValue(books, forKey: book.author)
                    break
                }
            }

            if books.count == 0 {
                return self.removeAuthor(author: book.author)
            }
        }

        return false
    }

    /// Update the book.
    ///
    /// - Parameter oldBook: New book data.
    /// - Parameter newBook: Old book data.
    /// - Returns: "true" if successful.
    func update(oldBook: Book, newBook: Book) -> Bool {
        if oldBook.author == newBook.author {
            return self.replaceBook(newBook: newBook)
        } else if self.remove(book: oldBook) {
            return self.add(book: newBook)
        }

        return false
    }

    /// Remove the author at the inner collection.
    ///
    /// - Parameter author: Name of the author.
    /// - Returns: "true" if successful.
    private func removeAuthor(author: String) -> Bool {
        self.booksByAuthor.removeValue(forKey: author)
        for i in 0..<self.authors.count {
            let existAuthor = self.authors[i]
            if existAuthor == author {
                self.authors.remove(at: i)
                return true
            }
        }

        return false
    }

    /// Replace the book.
    ///
    /// - Parameter newBook: New book data.
    /// - Returns: "true" if successful.
    private func replaceBook(newBook: Book) -> Bool {
        if var books = self.booksByAuthor[newBook.author] {
            for i in 0..<books.count {
                let book = books[i]
                if book.bookId == newBook.bookId {
                    books[i] = newBook
                    self.booksByAuthor.updateValue(books, forKey: newBook.author)
                    return true
                }
            }
        }

        return false
    }
}
